common/src/main/java/org/conscrypt/OpenSSLAeadCipher.java

```diff
@@ -220,6 +220,55 @@ boolean allowsNonceReuse() {
         return false;
     }
 
+    /**
+     * Encrypts of decrypts data in a single-part operations or finishes multiple part operation.
+     * @param input Input data that is returned fully read
+     * @param output the storage buffer of the resulting encryption decryption. Output is returned position
+     *              incremented by bytesWritten Limit stays the same
+     * @return number of bytes the output buffer's position has moved by.
+     * @throws ShortBufferException if output.remaining() bytes are insufficient to hold the result.
+     * @throws IllegalBlockSizeException
+     * @throws BadPaddingException
+     */
+    @Override
+    protected int engineDoFinal(ByteBuffer input, ByteBuffer output) throws ShortBufferException,
+            IllegalBlockSizeException, BadPaddingException, IllegalArgumentException {
+        if (input == null || output == null) {
+            throw new NullPointerException("Null ByteBuffer Error");
+        }
+        if (getOutputSizeForFinal(input.remaining()) > output.remaining()) { // uses the child class's size if available
+            throw new ShortBufferWithoutStackTraceException("Insufficient Bytes for Output Buffer");
+        }
+        if (output.isReadOnly()) {
+            throw new IllegalArgumentException("Cannot write to Read Only ByteBuffer");
+        }
+        if (bufCount != 0) {
+            return super.engineDoFinal(input, output);// traditional case
+        }
+        int bytesWritten;
+        if (!input.isDirect()) {
+            int incap = input.remaining();
+            ByteBuffer inputClone = ByteBuffer.allocateDirect(incap);
```

Comment:
Not this CL, but I wonder if its worth keeping a small cache of these allocations as `allocateDirect` isn't cheap.  Probably more of an issue for `engineDoUpdate` than here though.